quantity/core.py

```diff
@@ -0,0 +1,243 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import annotations
+
+import operator
+from dataclasses import dataclass, replace
+from typing import TYPE_CHECKING
+
+import array_api_compat
+import astropy.units as u
+import numpy as np
+from astropy.units.quantity_helper.converters import UFUNC_HELPERS
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+DIMENSIONLESS = u.dimensionless_unscaled
+
+PYTHON_NUMBER = float | int | complex
+
+
+def has_array_namespace(arg):
+    try:
+        array_api_compat.array_namespace(arg)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def get_value_and_unit(arg, default_unit=None):
+    # HACK: interoperability with astropy Quantity.  Have protocol?
+    try:
+        unit = arg.unit
+    except AttributeError:
+        return arg, default_unit
+    else:
+        return arg.value, unit
+
+
+@dataclass(frozen=True, eq=False)
+class Quantity:
+    value: Any
```

Comment:
Should this be "stuff that implements the Array API". 
Unfortunately the Array API doesn't vendor a Protocol themself, but based on the docs the minimal protocol would be 

```python
class SupportsArrayAPI(Protocol):
    def __array_namespace__(self, *, api_version: str | None = None) -> Any: ...
```

```suggestion
    value: SupportsArrayAPI
```

Reply:
The problem is that apart from numpy, no-one actually defines `__array_namespace__` (well, at least `dask` and `jax` do not). Also, I do want to add support for python numbers. Overall, it seems easier to just punt on this for now.

Reply:
Jax should define `__array_namespace__` ([link](https://github.com/jax-ml/jax/blob/c5dc980db80a55fd2a0583208921f84a854dfc97/docs/jax.numpy.rst#L548).
Hopefully dask will soon as well.
IMO supporting python numbers should be handled by a wrapper class that provides the Array API.

```python
import math

class PythonNumber:
    def __init__(self, value): ...
    def __array_namespace__(self, ...): return math
```

If we stick to the Array API it will massively reduce compatibility logic in every function..

Reply:
Ah, you're right, JAX actually does define it already.

For python numbers, I was thinking along similar lines. That may be where your `__value_namespace__` can come in handy. Although perhaps wrapping `array_api_compat.array_namespace` may be better. 

Anyway, how about we do a typing PR after this one? I do think it is a good idea to start with this as well-typed as possible, but fear I'll need more than a little help!

Reply:
> Although perhaps wrapping array_api_compat.array_namespace may be better.

Agreed. If we can avoid `__value_namespace__` entirely, that would be nice.